openssl/src/ssl/mod.rs

```diff
@@ -297,6 +303,54 @@ extern fn raw_verify_with_data<T>(preverify_ok: c_int,
     }
 }
 
+extern fn raw_sni(ssl: *mut ffi::SSL, ad: &mut c_int, arg: *mut c_void)
+        -> c_int {
+    unsafe {
+        let ssl_ctx = ffi::SSL_get_SSL_CTX(ssl);
+        let callback = ffi::SSL_CTX_get_ex_data(ssl_ctx, SNI_IDX);
+        let callback: Option<ServerNameCallback> = mem::transmute(callback);
+        let mut s = Ssl { ssl: ssl };
+
+        let res = match callback {
+            None => ffi::SSL_TLSEXT_ERR_ALERT_FATAL,
+            Some(callback) => callback(&mut s, ad)
+        };
+
+        // Allows dropping the Ssl instance without calling SSL_FREE on the SSL object
+        s.ssl = ptr::null_mut() as *mut ffi::SSL;
```

Comment:
Following our conversation on IRC, I was wary of potential memory leaks using `mem::forget` here, but I just ran a test for this, and I see no evidence of a leak, so let's go with that solution.